source/extensions/quic_listeners/quiche/codec_impl.h

```diff
@@ -0,0 +1,58 @@
+#include "envoy/http/codec.h"
+
+#include "common/common/assert.h"
+#include "common/common/logger.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_server_session.h"
+
+namespace Envoy {
+namespace Quic {
+
+// QuicHttpConnectionImplBase instance is a thin QUIC codec just providing quic interface to HCM.
+// Owned by HCM and created during onNewConnection() if the network connection
+// is a QUIC connection.
+class QuicHttpConnectionImplBase : public virtual Http::Connection,
+                                   protected Logger::Loggable<Logger::Id::quic> {
+public:
+  QuicHttpConnectionImplBase(EnvoyQuicServerSession& quic_session) : quic_session_(quic_session) {}
+
+  // Http::Connection
+  void dispatch(Buffer::Instance& /*data*/) override {
+    // Bypassed. QUIC connection already hands all data to streams.
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+  Http::Protocol protocol() override {
+    // From HCM's view, QUIC should behave the same as Http2, only the stats
+    // should be different.
+    // TODO(danzh) add Http3 enum value for QUIC.
```

Comment:
Q for my own understanding: Would we actually do this? Is HTTP/3 different from HTTP/2 in any substantial way at the app layer? I'm mainly wondering if it's worth it to make Envoy aware of HTTP/3 above the codec.

Reply:
Distinguish Http3 from Http2 traffic is more for monitoring. These two works the same at the app layer. The distinction should only exists in codec